fdbrpc/DDSketch.h

```diff
@@ -0,0 +1,229 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef DDSKETCH_H
+#define DDSKETCH_H
+#pragma once
+
+#include <vector>
+#include <algorithm>
+#include <cmath>
+#include "flow/Error.h"
+#include "flow/UnitTest.h"
+
+#if (!(__BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__))
+#error Do not support non-little-endian systems
+#endif
+
+#define FASTLOG 1
+// This significantly reduces the time for log
+
+#if FASTLOG
+namespace fastLogger {
+// Basically, the goal is to compute log(x)/log(r).
+// For double, it is represented as 2^e*(1+s) (0<=s<1), so our goal becomes e*log(2)/log(r)*log(1+s),
+// and we approximate log(1+s) with a cubic function. See more details on Datadog's paper, or
+// CubicallyInterpolatedMapping.java in https://github.com/DataDog/sketches-java/
+inline const double correctingFactor = 1.00988652862227438516; // = 7 / (10 * log(2));
+constexpr inline const double A = 6.0 / 35.0, B = -3.0 / 5.0, C = 10.0 / 7.0;
+
+inline double fastlog(double value) {
+	int e;
+	double s = frexp(value, &e);
+	s = s * 2 - 1;
+	return ((A * s + B) * s + C) * s + e - 1;
+}
+
+inline double reverseLog(double index) {
+	long exponent = floor(index);
+	// Derived from Cardano's formula
+	double d0 = B * B - 3 * A * C;
+	double d1 = 2 * B * B * B - 9 * A * B * C - 27 * A * A * (index - exponent);
+	double p = cbrt((d1 - sqrt(d1 * d1 - 4 * d0 * d0 * d0)) / 2);
+	double significandPlusOne = -(B + p + d0 / p) / (3 * A) + 1;
+	return ldexp(significandPlusOne / 2, exponent + 1);
+}
+}; // namespace fastLogger
+#endif
+
+// A DDSketch for non-negative numbers
+// (those < 10^-18 are treated as 0, and huge numbers (>>10^18) fail ASSERT)
+template <class T>
+class DDSketch {
+public:
+	explicit DDSketch(double errorGuarantee = 0.1)
+	  : errorGuarantee(errorGuarantee), gamma((1.0 + errorGuarantee) / (1.0 - errorGuarantee)), logGamma(log(gamma)),
+	    multiplier(FASTLOG ? fastLogger::correctingFactor * log(2) / log(1 + 2 * errorGuarantee / (1 - errorGuarantee))
+	                       : 1),
+	    offset((FASTLOG ? fastLogger::fastlog(1 / EPS) * multiplier : log(1 / EPS)) + 5), populationSize(0),
+	    zeroPopulationSize(0), minValue(T()), maxValue(T()), sum(T()) {
+		buckets.resize(2 * offset, 0);
+	}
+
+	DDSketch<T>& addSample(T sample) {
+		// Call it addSample for now, while it is not a sample anymore
+		if (!populationSize) minValue = maxValue = sample;
+
+		if (sample <= EPS) {
+			zeroPopulationSize++;
+		} else {
+#if FASTLOG
+			int index = ceil(fastLogger::fastlog(sample) * multiplier);
+#else
+			int index = ceil(log(sample) / logGamma);
```

Comment:
log(gamma) ~= .25 with default parameters, so this is ~= ceil(log(sample) * 4.0)  That means we don't need much precision out of log() here either.  (So, either mode could be reduced to the clz trick in my other comment)

Reply:
I'm not sure whether I understand the suggested alternative method correctly, but it looks to me that `intlog((sample - 1) << 4) + 1` (or a similar approach) computes  `ceil(log(sample)) + 4` instead of `ceil(log(sample) * 4)`.

The difference is critical. Just imagine this, all numbers between `2^10` and `2^11` (ignore any potential off-by-1) have the same value 10 after `intlog`. They are treated as the same and will be inserted into the same bucket. The relative error here is too big (potential `sqrt(2)` which is around 40%). What the original commits try to do is to divide numbers in this range into 4 (or 3.4, which is actually 5 because ln(1.1/0.9) is 0.200, but the actual number does not really matter) buckets, specifically `[2^10, 2^10.25)`, `[2^10.25, 2^10.5)`, `[2^10.5, 2^10.75)`, `[2^10.75, 2^11)`. And we can find exactly where a certain percentage is located at. So, a float-number `log` is important here (and we will 'amplify' it before taking the `ceil`), even if all inputs are integer numbers.

Reply:
Whoops.  That's right.  We need to compute `sample^4`, not `sample * 2^4`.  For a 32 bit sample, a uint128 is long enough to hold the intermediate value.  Fortunately, uint128's are supported by g++ and clang.

They seem to be invoking a 128bit unsigned multiply instruction.  If I just want the first 32 bits of the result, it emits this:
```asm
main:                                   # @main
        movsxd  rdx, edi
        imul    rdx, rdx
        mulx    rax, rax, rdx
        shr     rax, 32
        ret
```
If I'm reading the documentation correctly, `mulx dest_hi, dest_lo, rdx` is computing `(dest_hi, dest_lo) = rdx * rdx`.  Assuming the emitted code works, it must discard dest_lo when you pass the same register in for both destination registers.

Here's an example that actually computes the clz of the result using the 64-bit `__builtin_clzll`:

```c++
#include <cstdint>

inline __uint128_t square(uint64_t v) {
    __uint128_t t = v;
    return t * t;
}

int main(int argc, char ** argv) {

    auto x = square(((uint64_t)argc) * argc);

    // This seems to generate just the code emitted by argc^4.
    // return x >> 96;

    uint64_t a, b;

    a = (uint64_t)x;
    b = (uint64_t)(x >> 64);

    return (!!a) * (__builtin_clzll(a) ) +
           (!!b) * (__builtin_clzll(b) + 64);
}
```
which emits:
```asm
main:                                   # @main
        movsxd  rdx, edi
        imul    rdx, rdx
        mulx    rcx, rax, rdx
        xor     edx, edx
        lzcnt   rsi, rax
        cmovb   esi, edx
        lzcnt   rax, rcx
        or      eax, 64
        test    rcx, rcx
        cmove   eax, edx
        add     eax, esi
        ret
```

I didn't check to see if there's a 128 bit lzcnt instruction; I'm not willing to write + maintain assembly for this.

This has more variants, and things that use __builtin_clz:

https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAM1QDsCBlZAQwBtMQBGAFlICsupVs1qhkAUgBMAISnTSAZ0ztkBPHUqZa6AMKpWAVwC2tEJNJb0AGTy1MAOWMAjTMRABWAAykADqgWE6rR6hiaCfgFqdDZ2jkYubpyKypiqQQwEzMQEIcam5koqUbQZWQQxDs6uHt4Kmdm5YUl1ZRVxCVwAlIqoBsTIHADkUgDMtsiGWADU4iM6yHX49LPY4p4Agmvrtqy2mFMA%2BgcGtgSckgAcBwRTCgCOBlmYECf0AGzc11MAbp0zAOyyDZTEGHY6nc5XG43WYAER%2BsyB61BU2ImAIfVoUxuACpsYituJ/rDCRtTlMjMxbBByVlgMhSFNkAgslMcXi6b8AUitijmAYiFMAB4zEbw%2B6PNFQF6nD7XTp05B/DnEemdAkbXmggD02qmABUEHgFLdMJgjCbBcAtK5mAR9nwDHVsQh9mhpubCPb0FMnABPKaK8TubDcAB0WpButR6MxwtFqxG2CmAE43hrNsDQa8CHKbsxGU4M5HA6L4TL3p8CJ0hRmUU4y1MK7mq50ICKVisph91SMeVmQWiMcQsRAwOOwMxlU2jk4Tqw1LQDhMAF6sVgQKdTP5yEso0FjidOacQWfzxfL1hrjfHmYybvcXtImYDqbRnMjSRfAu%2BxkMqboHWOp6j6cJNh%2BX7VrWfYltGHZJp2YpTJ%2BQFRnqyCNs2n7ytBz7AfGiGJvCKEwa%2B0YNmBWGQTWqFvnq8GJkRyGSLR0bMJhEE4cWZF6osIAoL0MJzLMOjgbKrbtgmXY9qKOgibcBDoPxlisKxvGKfxaACrJ8nNnmbYMdJj7cci%2BGGpgSjYgA7qgMbDliLT2kYWgECaCgIL0rA%2Bi4LLfPsBCurczDOYiLr7FQeDEM6nqub6xAiMyFlTFZhAIEywiiIyAX7EoaDaABqAWbQYCDAQEY8VMACSUy%2BVoJU3C4tDMpSxAANaYD6gpKPsVlGsyUzGlMVDMHUrjlZmploXZmKwXqILsbMABiYJzngC62JeK6bn8ICzaCFEjMtZ5rReq4QLecjMVMu0VSCGFLSt54bWdSp3tID7XXtIKgYdj0nc9V4QOgO73mmn23VMngmSi0ZDjNEMTpObIzgcq3rUuZ1biiIN4VNIKIw2eKnqjT0Y4DF33p%2BONfVMiMYUTx3o5tECvZdMm7gjE4%2BgzJP/WTW3A29qZvE%2BhLEoM3SsCAgzuIMpCmIMnhy6g0tyTIl0KL0/T7KMnBywQ0tK503StSA3D/GGKacJwbwptw7icJ4/wjCmFz/EI0vcHLCtK6QKuDHLCggN4BuKxLpBwLAMCICAgk%2BAK5CUGgRg%2BGt1RJB1hAkIIND0EwbAcDwpBWfFPjS3rkvS7L8uG8r0u68lqXChcbwALQfOlIjAFM1thp4ff67XxukK6zBYG4ECV4MXukEYZgW473Bu9wK9vJ4FzuG87s%2B3XAeKMHpCh0bEfRxASDJ6n7BkBQLOoCnaduBMXfW543gRQurhB%2BdtekE4thZH6cuctk7OXoAAeVoKwQBYdSBYEpKIdgP98BojSH5IOMDMBClSAKIYetyD0GUD/XYTh4rED9HoLAP8CDEDwHPQYFdSC50YCwRBRcBBJAymIXcMghB4CcEHWAtBgocCoEYAgpA/JuEGG8SQFduioB8MUdBrdFhwgkOrGQkh2KtzASMKYrcWrMjhEItQflA4pDSBoCAlhGimAztoNoVREi%2BH8IEOgtjwiuOKI4%2BI6dkhFHSPUHI%2Bg8iCEKKkYopRsg%2BI6M0IJHi4mtD2O0dO3RNZ9AGF0D2MtvY/39kKFu7duCd1ED3Tgfc%2B5NlwFnYgd4RhJCmHoe%2BV86mcD%2BGrWQMhB5h2HqbT8fc3j1Jtm8N47h3D/G4G8Dhns5Zz3%2BJIMMFwUwO0kGM9wLseAjBGDXGB/tA4HyPhLE2HgFnrwuCMTw7gl4pkkGvG22SRi5N2dLbpRsp6SCeb7PZh8h7dEkQEDQ3AgA

Here's an article where someone implemented uint128_t squares in hand using c#; it might be possible to backport to java:

https://www.codeproject.com/tips/618570/uint-multiplication-squaring

There's also a 256 bit variant, in case we want to be able to support 64 bit samples.

Reply:
Got it. This is a nice proposal, I will try to keep two versions, possibly implemented in CRTP (please let me know if this is a bad idea), `DDSketch` and `DDSketchFastInt`.

For now `DDSketchFastInt` will support around `13.5%` guaranteed relative accuracy (by taking the power of 4) and only support `unsigned int`.

Reply:
Thanks!  That'll let us unify everything around one type of sketch.

I've been meaning to change things so that all latencies are represented using uint32_t's, which will let us move pretty much everything to the fast variant.  That should be a in a different PR.